baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/IllegalSafeLoggingArgument.java

```diff
@@ -497,4 +499,67 @@ private Safety getLambdaRequiredReturnSafety(LambdaExpressionTree tree, VisitorS
         }
         return SafetyAnnotations.getSafety(returnType.type(), state);
     }
+
+    @Override
+    public Description matchMemberReference(MemberReferenceTree tree, VisitorState state) {
+        // This is the type the reference gets "cast" into, whether through assignment, return type, argument, etc
+        Type castType = state.getTypes().findDescriptorType(ASTHelpers.getType(tree));
+        // The reference will be used as the expected type. This means:
+        //   * the safety of the return type of the expected must be "lower" than the actual safety of the reference
+        //   * the safety of the arguments of the expected must be "higher" than the actual safety of the reference
+        // i.e. "Supplier<@Safe String> f = this::method" shouldn't be allowed, when method returns unsafe
+        // Similarly, "@Consumer<@Unsafe String> f = this::method" shouldn't be allowed, when method expects to take a
+        //   safe argument (since it might log it)
+
+        Safety castTypeReturnSafety = SafetyAnnotations.getSafety(castType.getReturnType(), state);
```

Comment:
should we check `allowsAll()` here to avoid checking the method symbol/returnType below in the `UNKNOWN` and `DO_NOT_LOG` cases?